crates/matrix-sdk-crypto/src/olm/account.rs

```diff
@@ -1502,12 +1506,52 @@ impl Account {
                     )
                     .into());
                 }
+                sender_device = Some(device);
             }
 
+            let verification_state = sender_device
+                .as_ref()
+                .map(|device| {
+                    if device.is_verified() {
+                        // The device is locally verified or signed by a verified user
+                        VerificationState::Verified
+                    } else if device.is_cross_signed_by_owner() {
+                        // The device is not verified, but it is signed by its owner
+                        if device
+                            .device_owner_identity
+                            .as_ref()
+                            .expect(
+                                "A device cross-signed by the owner must have an owner identity",
+                            )
+                            .was_previously_verified()
+                        {
+                            VerificationState::Unverified(VerificationLevel::VerificationViolation)
+                        } else {
+                            VerificationState::Unverified(VerificationLevel::UnverifiedIdentity)
+                        }
+                    } else {
+                        // No identity or not signed
+                        VerificationState::Unverified(VerificationLevel::UnsignedDevice)
+                    }
+                })
+                .unwrap_or(VerificationState::Unverified(VerificationLevel::None(
+                    DeviceLinkProblem::MissingDevice,
+                )));
+
+            let encryption_info = EncryptionInfo {
+                sender: event.sender().to_owned(),
+                sender_device: sender_device.map(|d| d.device_id().to_owned()),
+                algorithm_info: AlgorithmInfo::OlmV1Curve25519AesSha2 {
+                    curve25519_public_key_base64: sender_key.to_base64(),
+                },
+                verification_state,
+            };
```

Comment:
Shame that this can't be a method on the `Device` but let's at least put this into a separate method in the `Account`.